### ISSUE TITLE
CSQA dataset and Dialog-to-Action model

### ISSUE THREAD

**Issue by rubenseggers** [allenai/allennlp#2472]
Hi all,

I'm working on a project with a group and we want to use AllenNLP. We have been working on extending the [MAPO paper](https://arxiv.org/abs/1807.02322), and our next goal is to reimplement an existing model on a graph based dataset in a more extendable framework.

Therefore, we would like to contribute to the addition of the the Complex Sequential Question Answering dataset: [CSQA](https://amritasaha1812.github.io/CSQA/). This dataset provides the task of i) question answering over a large knowledge base, and ii) sequential question answering in the sense that previous questions can be used as context to provide proper answers for follow-up questions. 

Furthermore, the paper [Dialog-to-Action: Conversational Question Answering Over a Large-Scale Knowledge Base ](https://papers.nips.cc/paper/7558-dialog-to-action-conversational-question-answering-over-a-large-scale-knowledge-base) is to the best of my knowledge the only paper to improve over the baseline results. Therefor, I believe it would be useful to add this model to AllenNLP. 

This would entail implementing:
- Data readers for the conversations;
- Embedders for entities;
- Transition function with memory component;
- MML trainer with loss function defined in Dialog-to-Action;
- Model wrapper combining all necessary components;
- Wikidata context class containing the knowledge base;
- State class containing the specific grammar statelet and RNN statelet;
- Grammar and execution engine (maybe some components in AllenNLP are already usable for this);
- World class around the context grammar and executor. 

We would love to discuss this contribution; Is there a communication channel besides GitHub for this purpose?

Thanks!
Ruben

**Comment by matt-gardner** [comment]
Hi Ruben,

I'd be thrilled to see this implemented in AllenNLP.  Probably the easiest way to communicate for in depth questions like this is to use the #allennlp channel in the pytorch slack.  Be sure to @ mention me so I see it - I don't check that slack often, unless I get a direct message.

One thing to note - I've recently re-written how the grammar / executor / world stuff works, so it is now _way_ simpler than it used to be.  We're still testing this out with a few more models, so we haven't officially "released" it yet or written a tutorial for it, but it should work.  The best place to start looking into this is by reading the docstrings [here](https://github.com/allenai/allennlp/blob/08a8c5e56d8728c17eb96de3ae603ce6799a5b2f/allennlp/semparse/domain_languages/domain_language.py#L220-L287), and then looking at some examples [here](https://github.com/allenai/allennlp/blob/08a8c5e56d8728c17eb96de3ae603ce6799a5b2f/allennlp/tests/semparse/domain_languages/domain_language_test.py#L10-L97) and [here](https://github.com/allenai/allennlp/tree/master/allennlp/semparse/domain_languages).  Note that this new `DomainLanguage` only works if the language you're implementing is purely functional.  I looked briefly at the paper you mentioned, and it looks like the language there is indeed functional, so you should be ok.

I'd recommend getting a first version working using a single turn in the dialog (i.e., just a standard semantic parsing task).  You should be able to do this using components that are already built in to the library, without needing to write your own transition function, state, or that kind of stuff.  You should just need a dataset reader, a domain language, and a model patterned very closely on something like the NLVR parser.  You're right that getting the KB embedded and used correctly in the model will be a big difference with NLVR, though there are some similarities with how the WikiTableQuestions parser works that you might be able to use.

After you have that version working, you can figure out how to add dialog state to it.

**Comment by koenvanderveen** [comment]
Hi Matt,

Thanks for you quick response, we really appreciate it. I am working with Ruben on this project, and as we do not have access to the pytorch slack yet (requested access today), I would like to ask another question here. Does the DomainLanguage class totally replace grammar / executor / world instances? If we do not have a world, where do track state?

We worked on the datasetreader class for CSQA + tests today, we will keep you posted on our progress.

**Comment by matt-gardner** [comment]
Yes, the `DomainLanguage` replaces all three of those (the `type_declaration` / grammar, the `executor`, and the `world`), because you define an executor (that gets initialized with whatever instance-specific state it needs - see the WikiTables language for an example of this), and we infer the grammar from the executor.  In my quick look at the paper, it looked like you didn't have any instance-specific state, so I recommended starting from the NLVR parser.  Is there something instance-specific that I missed?

The decoding / conversation state is handled totally separately (in the `GrammarStatelet` class, among others).

**Comment by matt-gardner** [comment]
Closing this issue as I've answered most of the questions here.
